module/zfs/dbuf.c

```diff
@@ -3874,6 +3878,7 @@ dbuf_hold_copy(dnode_t *dn, dmu_buf_impl_t *db)
 		    DBUF_GET_BUFC_TYPE(db), db->db.db_size));
 	}
 
+	ASSERT(MUTEX_HELD(&db->db_mtx));
 	rw_enter(&db->db_rwlock, RW_WRITER);
 	memcpy(db->db.db_data, data->b_data, arc_buf_size(data));
 	rw_exit(&db->db_rwlock);
```

Comment:
This looks odd to me.  We first allocated empty buffer and assigned it to `db->db.db_data` and only then after taking lock filling it with data.  I don't feel deep enough in `db_rwlock` semantics, and I am not saying this is necessary a problem, but I would probably feel better if we move `dbuf_set_data()` after the filling the allocated buffer.

Reply:
That makes sense.  Then we won't need the db_rwlock.